### ISSUE TITLE
Understanding the trainer parameters

### ISSUE THREAD

**Issue by vahuja4** [huggingface/setfit#364]
I am looking at the SetFit example with SetFitHead:
```
# Create trainer
trainer = SetFitTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    loss_class=CosineSimilarityLoss,
    metric="accuracy",
    batch_size=16,
    num_iterations=20, # The number of text pairs to generate for contrastive learning
    num_epochs=1, # The number of epochs to use for contrastive learning
    column_mapping={"sentence": "text", "label": "label"} # Map dataset columns to text/label expected by trainer
)


```
Here, what exactly is the meaning of `num_iterations`?  And, why is `num_epochs =1`? Is that sufficient?

```
trainer.unfreeze(keep_body_frozen=False)

trainer.train(
    num_epochs=25, # The number of epochs to train the head or the whole model (body and head)
    batch_size=16,
    body_learning_rate=1e-5, # The body's learning rate
    learning_rate=1e-2, # The head's learning rate
    l2_weight=0.0, # Weight decay on **both** the body and head. If `None`, will use 0.01.
)
metrics = trainer.evaluate()
```
Here the number of epochs is 25 and we are training the head and the body with different learning rates. Any reason why the per-epoch metrics are not displayed?

**Comment by tomaarsen** [comment]
> Here, what exactly is the meaning of num_iterations? And, why is num_epochs =1? Is that sufficient?

`num_iterations` is a somewhat poorly named parameter that indicates how many times we create "contrastive pairs". Every time we create these contrastive pairs, we create one "positive pair" (two sentences that are of the same class) and one "negative pair" (two sentences that are of another class) for every sentence. So, 2 pairs per sentence every time that we create the contrastive pairs. If we do it `num_iterations` times, then we create `num_sentences * 2 * num_iterations` pairs in total. 

Doubling the `num_iterations` is roughly equivalent to doubling `num_epochs`, with one notable difference: with `num_iterations` being doubled, we "freshly" generate more pairs for training, while with doubling `num_epochs` we reuse the same pairs we already generated, if that makes sense. In practice, the former is recommended, so we tend to recommend increasing `num_iterations` rather than increasing `num_epochs` when you want to train for longer.

The above counts specifically for finetuning the underlying Sentence Transformer. That is step 1 of SetFit. Step 2 is training the classifier, which converts the embeddings produced by the underlying sentence transformers into the correct class labels. When using the `SetFitHead`, this is roughly "traditional" training. With other words, no fancy pairs. This means that we now have `num_sentences` samples (instead of `num_sentences * 2 * num_iterations`), so here we have to increase the batch number of epochs to get the model to really learn this.

We don't have per-epoch metrics because there is no mid-training evaluation or logging implemented. Ideally, that would become a part of the upcoming v1.0.0 release.

I hope this clarifies things a bit.

- Tom Aarsen
